javascript/ql/src/semmle/javascript/security/UselessUseOfCat.qll

```diff
@@ -0,0 +1,320 @@
+  SystemCommandExecution command;
+
+  CommandCall() { this = command }
+
+  /**
+   * Holds if the call is synchronous (e.g. `execFileSync`).
+   */
+  predicate isSync() { command.isSync() }
+
+  /**
+   * Gets an argument to this command execution that specifies the argument list to the command.
+   */
+  DataFlow::Node getArgumentList() { result = command.getArgumentList() }
+
+  /**
+   * Gets the callback (if it exists) for an async `exec`-like call.
+   */
+  DataFlow::FunctionNode getCallback() {
+    not this.isSync() and result = getLastArgument().getALocalSource()
+  }
+
+  /**
+   * Holds if the executed command execution has an argument list as a separate argument.
+   */
+  predicate hasArgumentList() { exists(command.getArgumentList()) }
+
+  /**
+   * Gets the data-flow node (if it exists) for an options argument for an `exec`-like call.
+   */
+  DataFlow::Node getOptionsArg() { result = command.getOptionsArg() }
+
+  /**
+   * Gets the constant-string parts that are not part of the command itself.
+   * E.g. for a command execution `exec("/bin/cat foo bar")` this predicate will have result `"foo bar"`.
+   */
+  string getNonCommandConstantString() {
+    if this.hasArgumentList()
+    then
+      result =
+        getConstantStringParts(getArgumentList()
+              .getALocalSource()
+              .(DataFlow::ArrayCreationNode)
+              .getElement(_))
+    else
+      exists(string commandString | commandString = getConstantStringParts(getArgument(0)) |
+        result = commandString.suffix(1 + commandString.indexOf(" ", 0, 0))
+      )
+  }
+
+  /**
+   * Holds if this command execution invokes the executeable `name`.
+   */
+  bindingset[name]
+  predicate isACallTo(string name) {
+    if this.hasArgumentList()
+    then getArgument(0).mayHaveStringValue(name)
+    else
+      exists(string arg | arg = getConstantStringParts(getArgument(0)) |
+        arg.prefix(name.length()) = name
+      )
+  }
+}
+
+/**
+ * Gets the constant string parts from a data-flow node.
+ * Either the result is a constant string value that the node can hold, or the node is a string-concatenation and the result is the string parts from the concatenation.
+ */
+private string getConstantStringParts(DataFlow::Node node) {
+  node.mayHaveStringValue(result)
+  or
+  result = node.(StringOps::ConcatenationRoot).getConstantStringParts()
+}
+
+/**
+ * A call to a useless use of `cat`.
+ */
+class UselessCat extends CommandCall {
+  UselessCat() {
+    this = command and
+    isACallTo(getACatExecuteable()) and
+    // There is a file to read, it's not just spawning `cat`.
+    not (
+      not exists(getArgumentList()) and
+      getArgument(0).mayHaveStringValue(getACatExecuteable())
+    ) and
+    // wildcards, pipes, redirections, other bash features, and multiple files (spaces) are OK.
+    not exists(getNonCommandConstantString().regexpFind("\\*|\\||>|<| |\\$|&|,|\\`", _, _)) and
```

Comment:
I think you dropped the space.
```suggestion
    not exists(getNonCommandConstantString().regexpFind("\\*|\\||>|<| |\\$|&|,|\\`| ", _, _)) and
```

Reply:
Can we factor out this check and apply it to the file-part of the alert message as a sanity check? It would be silly if we somehow suggested to replace something with `fs.readFile('foo>bar')`..

Reply:
> Can we factor out this check and apply it to the file-part of the alert message as a sanity check? It would be silly if we somehow suggested to replace something with fs.readFile('foo>bar')..

We can, but I really doubt that your example could happen.